### PR TITLE
Add Quota Validity Timer for UPG

### PR DESCRIPTION
**Pull request by sergeymatov**
- Added generic functionality
- Simple E2E test

### CODE REVIEW

File: test/e2e/framework/sessionconfig.go [hunk 1]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
...
+		for i, _ := range valTimer.Payload {
+			valTimer.Payload[i] = 0
+		}
+		valTimer.Payload[3] = 10
+
+	}
+	return urr
+}
+
+func (cfg SessionConfig) SessionIEs(triggers uint16) []*ie.IE {
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:260]: It would be better to add `ReportingTriggers` field in `SessionConfig` (note that zero value is the default so there will be no need to set it). This structure is also used by `pfcpstub` in FPP test setup so it will have to be changed if the argument is added here.

File: test/e2e/framework/sessionconfig.go [hunk 2]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
...
-		ies = append(ies,
-			ie.NewCreateURR(
-				ie.NewURRID(1),
-				ie.NewMeasurementMethod(0, 1, 1), // VOLUM=1 DURAT=1
-				ie.NewReportingTriggers(0)),
-			ie.NewCreateURR(
-				ie.NewURRID(2),
-				ie.NewMeasurementMethod(0, 1, 1), // VOLUM=1 DURAT=1
-				ie.NewReportingTriggers(0)))
+			ies = append(ies, cfg.CreateURR(1, triggers), cfg.CreateURR(2,triggers))
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:263]: If `triggers` is replaced by `ReportingTriggers` as suggested above, there will be no need to pass it here. Also, please don't forget about `go fmt`, I believe it can be automated (do it on save) on most editors

File: upf/pfcp.c [hunk 3]
```diff
@@ -2364,6 +2366,9 @@ encode_volume_measurement (void *p, u8 ** vec)
 #define decode_quota_holding_time decode_u32_ie
 #define encode_quota_holding_time encode_u32_ie
 
+#define format_quota_validity_time format_u32_ie
+#define decode_quota_validity_time decode_u32_ie
+#define encode_quota_validity_time encode_u32_ie
```
Comment (RoadRunnr) [inline upf/pfcp.c:2371]: wrong IE order, this should go after alternative_smf_ip_address and before the TP IEs

File: test/e2e/framework/sessionconfig.go [hunk 4]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
...
+	// for Quota Validity Time report trigger is 15th bit in host order
+	if (triggers & 0x80) != 0 {
+		urr.Add(ie.NewQuotaValidityTime(time.Now()))
+
+		valTimer, _ := urr.FindByType(ie.QuotaValidityTime)
+		for i, _ := range valTimer.Payload {
+			valTimer.Payload[i] = 0
+		}
+		valTimer.Payload[3] = 10
+
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:255]: Remove blank line before `}`

File: upf/pfcp.c [hunk 5]
```diff
@@ -5099,6 +5104,10 @@ static struct pfcp_group_ie_def pfcp_create_urr_group[] =
       .type = PFCP_IE_QUOTA_HOLDING_TIME,
       .offset = offsetof(pfcp_create_urr_t, quota_holding_time)
     },
+    [CREATE_URR_QUOTA_VALIDITY_TIME] = {
+      .type = PFCP_IE_QUOTA_VALIDITY_TIME,
+      .offset = offsetof(pfcp_create_urr_t, quota_validity_time)
+    },
```
Comment (RoadRunnr) [inline upf/pfcp.c]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: upf/pfcp.c [hunk 6]
```diff
@@ -5395,6 +5404,10 @@ static struct pfcp_group_ie_def pfcp_update_urr_group[] =
       .type = PFCP_IE_QUOTA_HOLDING_TIME,
       .offset = offsetof(pfcp_update_urr_t, quota_holding_time)
     },
+    [UPDATE_URR_QUOTA_VALIDITY_TIME] = {
+      .type = PFCP_IE_QUOTA_VALIDITY_TIME,
+      .offset = offsetof(pfcp_update_urr_t, quota_validity_time)
+    },
```
Comment (RoadRunnr) [inline upf/pfcp.c]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: upf/pfcp.h [hunk 7]
```diff
@@ -1674,6 +1679,7 @@ enum
   CREATE_URR_TIME_THRESHOLD,
   CREATE_URR_TIME_QUOTA,
   CREATE_URR_QUOTA_HOLDING_TIME,
+  CREATE_URR_QUOTA_VALIDITY_TIME,
```
Comment (RoadRunnr) [inline upf/pfcp.h]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: upf/pfcp.h [hunk 8]
```diff
@@ -1702,6 +1708,7 @@ typedef struct
   pfcp_time_threshold_t time_threshold;
   pfcp_time_quota_t time_quota;
   pfcp_quota_holding_time_t quota_holding_time;
+  pfcp_quota_validity_time_t quota_validity_time;
```
Comment (RoadRunnr) [inline upf/pfcp.h]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: upf/pfcp.h [hunk 9]
```diff
@@ -1727,6 +1734,7 @@ enum
   UPDATE_URR_TIME_THRESHOLD,
   UPDATE_URR_TIME_QUOTA,
   UPDATE_URR_QUOTA_HOLDING_TIME,
+  UPDATE_URR_QUOTA_VALIDITY_TIME,
```
Comment (RoadRunnr) [inline upf/pfcp.h]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: upf/pfcp.h [hunk 10]
```diff
@@ -1755,6 +1763,7 @@ typedef struct
   pfcp_time_threshold_t time_threshold;
   pfcp_time_quota_t time_quota;
   pfcp_quota_holding_time_t quota_holding_time;
+  pfcp_quota_validity_time_t quota_validity_time;
```
Comment (RoadRunnr) [inline upf/pfcp.h]: wrong order, Quota Validity Time should go after Dropped DL Traffic Threshold to keep it in sync with TS 29.244

File: test/e2e/framework/sessionconfig.go [hunk 11]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
 	}
 }
 
-func (cfg SessionConfig) SessionIEs() []*ie.IE {
+func (cfg SessionConfig) CreateURR (id uint32, triggers uint16) *ie.IE {
+	urr := ie.NewCreateURR(
+		ie.NewURRID(id),
+		ie.NewMeasurementMethod(0, 1, 1), // VOLUM=1 DURAT=1
+		ie.NewReportingTriggers(triggers))
+	// for Quota Validity Time report trigger is 15th bit in host order
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:246]: This comment is a bit out of place, better have a constant for that trigger

File: upf/upf.h [hunk 12]
```diff
@@ -596,6 +597,7 @@ typedef struct
   urr_time_t time_quota;	/* relative duration in seconds */
   urr_time_t quota_holding_time;	/* relative duration in seconds */
   urr_abs_time_t monitoring_time;	/* absolute UTC ts since 1900-01-01 00:00:00 */
+  urr_time_t quota_validity_time;	/* relative duration in seconds */
```
Comment (RoadRunnr) [inline upf/upf.h:600]: wrong order, Quota Validity Time should go before Monitoring Time to keep it in sync with TS 29.244.

Although it would also move before the urr_abs_time_t type just for readability.

File: test/e2e/framework/sessionconfig.go [hunk 13]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
...
+	if (triggers & 0x80) != 0 {
+		urr.Add(ie.NewQuotaValidityTime(time.Now()))
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:248]: If this is a dummy time value to be replaced later, you could probably just use `time.Time{}` instead of `time.Now()`

File: test/e2e/framework/sessionconfig.go [hunk 14]
```diff
@@ -237,18 +238,29 @@ func (cfg SessionConfig) DeleteFARs() []*ie.IE {
...
+
+		valTimer, _ := urr.FindByType(ie.QuotaValidityTime)
+		for i, _ := range valTimer.Payload {
+			valTimer.Payload[i] = 0
+		}
+		valTimer.Payload[3] = 10
```
Comment (ivan4th) [inline test/e2e/framework/sessionconfig.go:254]: As far as I understand, the problem is that Quota Validity Time in go-pfcp version we use wrongly uses `Time` instead of `Duration`? This should probably be noted in a comment above this part of code.
```
// FIXME: go-pfcp QuotaValidityTime definition is incorrect, as it should contain
// a Duration, not Time. Here we set it to be 10 seconds

File: test/e2e/upg_e2e.go [hunk 15]
```diff
@@ -534,6 +534,47 @@ var _ = ginkgo.Describe("Clearing message queue", func() {
...
+		f := framework.NewDefaultFramework(framework.UPGModeTDF, framework.UPGIPModeV4)
+		ginkgo.It("should generate usage report", func () {
+			ginkgo.By("Creating session with URR")
+			sessionCfg := &framework.SessionConfig{
+				IdBase: 1,
+				UEIP:   f.UEIP(),
+				Mode:   f.Mode,
+			}
+			reportCh := f.PFCP.AcquireReportCh()
+			triggers := 0x80
```
Comment (ivan4th) [inline test/e2e/upg_e2e.go:548]: Please avoid magic numbers in the code, introduce a constant.

File: upf/upf_pfcp.c [hunk 16]
```diff
@@ -1019,6 +1019,7 @@ pfcp_disable_session (upf_session_t * sx)
     upf_pfcp_session_stop_urr_time (&urr->time_threshold, now);
     upf_pfcp_session_stop_urr_time (&urr->time_quota, now);
     upf_pfcp_session_stop_urr_time (&urr->traffic_timer, now);
+    upf_pfcp_session_stop_urr_time (&urr->quota_validity_time, now);
```
Comment (RoadRunnr) [inline upf/upf_pfcp.c:1022]: I would put this after time_quota.

Side Note: it is somewhat strange that the quota_holding_time and monitoring_time is not cleared as well. But that would be a topic for another PR.

File: upf/upf_pfcp.c [hunk 17]
```diff
@@ -1980,6 +1989,7 @@ pfcp_update_apply (upf_session_t * sx)
 	    upf_pfcp_session_stop_urr_time (&urr->time_threshold, now);
 	    upf_pfcp_session_stop_urr_time (&urr->time_quota, now);
 	    upf_pfcp_session_stop_urr_time (&urr->traffic_timer, now);
+	    upf_pfcp_session_stop_urr_time (&urr->quota_validity_time, now);
```
Comment (RoadRunnr) [inline upf/upf_pfcp.c:1992]: I would put this after time_quota.

Side Note: it is somewhat strange that the quota_holding_time and monitoring_time is not cleared as well. But that would be a topic for another PR.

File: test/e2e/upg_e2e.go [hunk 18]
```diff
@@ -534,6 +534,47 @@ var _ = ginkgo.Describe("Clearing message queue", func() {
...
+			framework.Logf("Trigger %d", triggers)
```
Comment (ivan4th) [inline test/e2e/upg_e2e.go:549]: `0x%04x` would perhaps be more readable in the output than `%d`? (if this is really needed at all)

File: upf/upf_pfcp_api.c [hunk 19]
```diff
@@ -1755,7 +1756,8 @@ handle_create_urr (upf_session_t * sx, pfcp_create_urr_t * create_urr,
 
     create->measurement_period.handle =
       create->time_threshold.handle =
-      create->time_quota.handle = create->traffic_timer.handle = ~0;
+      create->time_quota.handle = create->traffic_timer.handle =
+      create->quota_validity_time.handle = ~0;
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:1760]: I would reorder this to have quota_validity_time directly after time_quota.

Side Note: it is somewhat strange that the quota_holding_time and monitoring_time is not initialized as well. But that would be a topic for another PR.

File: upf/upf_pfcp_api.c [hunk 20]
```diff
@@ -1801,6 +1803,12 @@ handle_create_urr (upf_session_t * sx, pfcp_create_urr_t * create_urr,
 	create->time_quota.period = urr->time_quota;
 	create->time_quota.base = now;
       }
+    if (ISSET_BIT (urr->grp.fields, CREATE_URR_QUOTA_VALIDITY_TIME))
+      {
+	create->update_flags |= PFCP_URR_UPDATE_QUOTA_VALIDITY_TIME;
+	create->quota_validity_time.period = urr->quota_validity_time;
+	create->quota_validity_time.base = now;
+      }
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:1811]: wrong order, Quota Validity Time should go after Dropped DL
Traffic Threshold  and before Monitoring Time to keep it in sync with TS 29.244.

**Review by ivan4th — CHANGES_REQUESTED** [review]

File: upf/upf_pfcp_api.c [hunk 21]
```diff
@@ -1908,6 +1916,12 @@ handle_update_urr (upf_session_t * sx, pfcp_update_urr_t * update_urr,
 	update->time_quota.period = urr->time_quota;
 	update->time_quota.base = update->start_time;
       }
+    if (ISSET_BIT (urr->grp.fields, UPDATE_URR_QUOTA_VALIDITY_TIME))
+      {
+	update->update_flags |= PFCP_URR_UPDATE_QUOTA_VALIDITY_TIME;
+	update->quota_validity_time.period = urr->quota_validity_time;
+	update->quota_validity_time.base = now;
+      }
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:1924]: wrong order, Quota Validity Time should go after Dropped DL
Traffic Threshold and before Monitoring Time to keep it in sync with TS 29.244.

File: test/e2e/upg_e2e.go (hunk 18 above)
Comment (sergeymatov) [reply test/e2e/upg_e2e.go:549]: forgot to remove :)

File: test/e2e/framework/sessionconfig.go (hunk 14 above)
Comment (sergeymatov) [reply test/e2e/framework/sessionconfig.go:254]: 1. It uses Time instead of Duration
2. It overrides provided Time value with some hardcode to make quota never expires. I will add a comment

File: upf/upf_pfcp.c (hunk 16 above)
Comment (sergeymatov) [reply upf/upf_pfcp.c:1022]: I believe UPG dont support holding/monitoring timers

File: test/e2e/framework/sessionconfig.go (hunk 1 above)
Comment (sergeymatov) [reply test/e2e/framework/sessionconfig.go]: Thanks for the note. Changed as a type field of `SessionConfig`

File: test/e2e/framework/sessionconfig.go (hunk 2 above)
Comment (sergeymatov) [reply test/e2e/framework/sessionconfig.go]: Done

File: test/e2e/framework/sessionconfig.go (hunk 4 above)
Comment (sergeymatov) [reply test/e2e/framework/sessionconfig.go]: Done

File: test/e2e/framework/sessionconfig.go (hunk 11 above)
Comment (sergeymatov) [reply test/e2e/framework/sessionconfig.go]: Done

File: upf/pfcp.c [hunk 22]
```diff
@@ -6569,6 +6582,7 @@ static struct pfcp_ie_def tgpp_specs[] =
     SIMPLE_IE(PFCP_IE_QUOTA_HOLDING_TIME, quota_holding_time, "Quota Holding Time"),
     SIMPLE_IE(PFCP_IE_DROPPED_DL_TRAFFIC_THRESHOLD, dropped_dl_traffic_threshold,
 	      "Dropped DL Traffic Threshold"),
+    SIMPLE_IE(PFCP_IE_QUOTA_VALIDITY_TIME, quota_validity_time, "Quota Validity Time"),
```
Comment (RoadRunnr) [inline upf/pfcp.c:6585]: wrong place, should go after PFCP_IE_ALTERNATIVE_SMF_IP_ADDRESS and before PFCP_IE_UE_IP_ADDRESS_POOL_INFORMATION

File: upf/upf_pfcp.c (hunk 16 above)
Comment (RoadRunnr) [reply upf/upf_pfcp.c:1021]: ahh, right, holding is not supported and monitoring is not a timer (its just a time)

**Review by ivan4th — APPROVED** [review]
LGTM

**Comment by sergeymatov** [comment]
@RoadRunnr are you fine with latest version of patch series?

**Review by RoadRunnr — APPROVED** [review]
C part LGTM
